katcp/resource_client.py

```diff
@@ -1114,16 +1124,37 @@ def wait(self, sensor_name, condition_or_value, timeout=5):
             sensor is not present
 
         """
+        if quorum is None:
+            quorum = len(self.clients)
+        elif quorum < 1:
+            quorum = int(math.ceil(quorum * len(self.clients)))
+        if timeout and max_grace_period:
+            # Avoid having a grace period longer than or equal to timeout
+            grace_period = min(max_grace_period, timeout / 2.)
+            initial_timeout = timeout - grace_period
+        else:
+            grace_period = max_grace_period
+            initial_timeout = timeout
         # Build dict of futures instead of list as this will be easier to debug
         futures = {}
         for client in self.clients:
-            futures[client.name] = client.wait(sensor_name, condition_or_value,
-                                               timeout)
-        results = yield futures
+            f = client.wait(sensor_name, condition_or_value, initial_timeout)
+            futures[client.name] = f
+        # No timeout required here as all futures will resolve after timeout
+        initial_results = yield until_some(done_at_least=quorum, **futures)
+        results = dict(initial_results)
+        # Identify stragglers and let them all respond within grace period
+        stragglers = {}
+        for client in self.clients:
+            if not results.get(client.name, False):
```

Comment:
That's actually not the same thing... I had it like that and then realised I want to revisit all clients with either False results or with no results. The former occurs in the midst of an initial timeout, where at least one future will resolve with False. These need to be revisited too.